docs/source/concepts/InstrumentAccessLayers.rst

```diff
@@ -31,216 +33,170 @@ ____________
+  if (!instrument)
+    throw runtime_error("There is no instrument in input workspace.")
 
-  if (!spectrumInfo.hasUniqueDetector(wsIndex)) {
-    // do exception
-    return false;
+  std::vector<IComponent_const_sptr> children;
+  instrument->getChildren(children, true);
+  std::vector<IComponent_const_sptr>::const_iterator it;
+  for (it = children.begin(); it != children.end(); ++it) {
+    if (const ObjComponent* obj = dynamic_cast<const ObjComponent*>(it->get())) {
+      // Do something with the obj component
+      obj.solidAngle(observer);
+    }
   }
 
-  // do stuff
-
-
-Access to Detectors
-___________________
-
-The ``detector(wsIndex)`` method on ``SpectrumInfo`` returns the parameterised detector or detector group for the workspace. As ``SpectrumInfo`` does not provide access to things like ``Object::solidAngle()``, the ``detector()`` method on ``SpectrumInfo`` can be used to get access to these methods.
-
-Mutable SpectrumInfo
-____________________
-
-The method ``MatrixWorkspace::mutableSpectrumInfo()`` returns a non-const ``SpectrumInfo`` object. Currently the only method that this access is required for is ``SpectrumInfo::setMasked(const size_t index, bool masked)``.
-
-Useful Tips
-___________
-
-* Creation of ``SpectrumInfo`` objects is not cheap. Make sure ``ws.spectrumInfo()`` is called outside of loops that go over all spectra.
-* If a ``SpectrumInfo`` object is required for more than one workspace then include the workspace name in the name of the ``SpectrumInfo`` object, to avoid confusion.
-* Get the ``SpectrumInfo`` object as a const reference and use auto - ``const auto &spectrumInfo = ws->spectrumInfo();``.
-* Do not forget to add the import - ``#include "MantidAPI/SpectrumInfo.h"``.
-
-Complete Examples
-_________________
-
-* `FindCenterOfMassPosition2.cpp <https://github.com/mantidproject/mantid/pull/17394/files#diff-905c244467474fc320eaf3b8c7a9f0dd>`_
-
-* `CreatePSDBleedMask.cpp <https://github.com/mantidproject/mantid/pull/18218/files#diff-f490acf06e76f93898dc7d486c8dfa93>`_
-
-* `HRPDSlabCanAbsorption.cpp <https://github.com/mantidproject/mantid/pull/18464/files#diff-fc151838d9d7cc2e4ea469e98472c791>`_
-
-DetectorInfo
-############
+**After - looping over index**
 
-Basics
-______
+.. code-block:: c++
 
-The conversion is similar to that for ``SpectrumInfo``. For ``DetectorInfo`` all instances of ``Instrument::getDetector()`` should be replaced using calls to the ``DetectorInfo`` object obtained from ``MatrixWorkspace::detectorInfo()``. The methods below can then be called on ``DetectorInfo`` instead of on the detector.
+  #include "MantidGeometry/Instrument/ComponentInfo.h"
 
-* ``isMonitor(detIndex)``
-* ``isMasked(detIndex)``
-* ``l2(detIndex)``
-* ``twoTheta(detIndex)``
-* ``signedTwoTheta(detIndex)``
-* ``position(detIndex)``
+  ...
 
-**Indexing**
+  const auto &componentInfo = ws->componentInfo();
+  if (componentInfo.size() == 0)
+    throw runtime_error("There is no instrument in input workspace.")
```

Comment:
See comment on old code, this would be adding an extra test that does not exist in the old code. Also the test is irrelevant for the code?